### PR TITLE
Fixed `ErrorCode`'s domain in Swift

### PR DESCRIPTION
**Pull request by NachoSoto**
## Background
In version 3, `Objective-C` had the constant `RCPurchasesErrorDomain` (`Purchases.ErrorDomain` in Swift), manually defined in Objective-C.
This was also visible to Swift through the constant of the same name, and both had the value `RCPurchasesErrorDomain`.

In version 4, written in Swift, it's not possible to expose global constants to Objective-C. The way the Swift compiler exposes error domains is through the definition of the `Error` itself.
In our case, it's `ErrorCode`, which has an `@objc` name `RCPurchasesErrorCode`. The compiler appends the word "Domain" to that, creating the new `RCPurchasesErrorCodeDomain` which superseded `RCPurchasesErrorDomain`.

However, prior to this change, our implementation of `CustomNSError` was not overriding `errorDomain`, so the value of these constants was the default `RevenueCat.ErrorCode` which was a change (regression) from version 3.x.
Unfortunately because of a bug (https://bugs.swift.org/browse/SR-15841), this isn't enough to fix the domain when `ErrorCode` is converted to `NSError`. This PR includes a workaround, but unfortunately the workaround only provides the correct value to Swift.
`RCPurchasesErrorCodeDomain` will still be the "invalid" `RevenueCat.ErrorCode`, but this is better than nothing.

At least now we have an explicit domain value for Swift errors, and we're exposing a constant for them (`Purchases.ErrorDomain` which was already available in version 3, so it's back now).

## TLDR:
In version 4 we had lost `Purchases.ErrorDomain` (which is arguably still useful to Swift), and the value was the auto-generated `RevenueCat.ErrorCode` instead of the one in version 3 `RCPurchasesErrorDomain`.
Swift now has a new custom domain `RevenueCatErrorDomain` which matches Apple's domains like `NSCocoaErrorDomain`.

### CODE REVIEW

**Review by joshdholtz — COMMENTED** [review]
giphy

**Review by aboedo — APPROVED** [review]
